source/extensions/tracers/datadog/event_scheduler.cc

```diff
@@ -0,0 +1,83 @@
+#include "source/extensions/tracers/datadog/event_scheduler.h"
+
+#include <memory>
+#include <utility>
+
+#include "source/common/common/assert.h"
+
+#include "datadog/json.hpp"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace Datadog {
+
+EventScheduler::EventScheduler(Event::Dispatcher& dispatcher) : dispatcher_(dispatcher) {}
+
+EventScheduler::~EventScheduler() {
+  for (const auto& timer : timers_) {
+    timer->disableTimer();
```

Comment:
nit: this is not needed. Timers will be disabled on destruction.

Reply:
I went digging to see whether that's true.

The interface's destructor doesn't mention it: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/envoy/event/timer.h#L30

This implementation's destructor is default, and so doesn't call `disableTimer` (more on this later): https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/timer_impl.h#L53

This implementation's destructor _does_ call `disableTimer`: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/scaled_range_timer_manager_impl.cc#L39

However, it looks like the scaled timer is distinct from "normal" timers in the dispatcher's implementation:
- https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/dispatcher_impl.cc#L67-L87
- https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/dispatcher_impl.cc#L213-L243

The code to follow seems to be: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/dispatcher_impl.cc#L237-L242

which is determined by the scheduler initialized here: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/dispatcher_impl.cc#L75

I'm guessing that scheduler ends up being the `RealScheduler`: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/real_time_system.cc#L16

In that case, it's the "base scheduler" that actually creates the timer. That's determined here: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/dispatcher_impl.h#L158

so the timer is created here: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/libevent_scheduler.cc#L38-L40

and that gives us one of these: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/timer_impl.h#L53

and its destructor does not call `disableTimer`.

However, looking at that implemention of `disableTimer`: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/timer_impl.cc#L30C17-L33

It deletes the underlying event. `TimerImpl` also derives from `ImplBase`: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/timer_impl.h#L53

`ImplBase` has a destructor that performs the same operation as `disableTimer`: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/source/common/event/event_impl_base.cc#L10

Aha!

Ok... my conclusion is: "If we assume `Event::Dispatcher` produces `Event::Timer` objects that are instances of `Event::TimerImpl`, then we may also assume that calling `disableTimer` is unnecessary on destruction, because an equivalent operation will be performed by one of the destructors of `Event::TimerImpl`."

Strictly speaking, the interface itself, `Event::Timer`, does not guarantee this behavior, and that is why I `disableTimer` in a loop in this destructor.

Do you still think it is best to remove this loop? I now agree that it's redundant in practice.

Reply:
I removed the loop, but still feel that it had its place.

Adding a comment to the `Timer` interface's destructor might be worth considering.